src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx

```diff
@@ -0,0 +1,803 @@
+import React, { useState } from 'react';
+import { connect } from 'react-redux';
+import _ from 'lodash';
+import { WithStyles, withStyles } from '@mui/styles';
+import {
+  Typography,
+  CustomTheme,
+  Button,
+  IconButton,
+  Menu,
+  MenuItem,
+  TextField,
+  Divider,
+  Autocomplete,
+  Checkbox,
+  FormControlLabel,
+} from '@mui/material';
+import {
+  AddCircleRounded,
+  IndeterminateCheckBoxOutlined,
+  AddBoxOutlined,
+  MoreVert,
+  Circle,
+} from '@mui/icons-material';
+import { TdrState } from 'reducers';
+import { useFormContext } from 'react-hook-form';
+import CodeMirror from '@uiw/react-codemirror';
+import { okaidia } from '@uiw/codemirror-theme-okaidia';
+import { javascript } from '@codemirror/lang-javascript';
+import {
+  ColumnModel,
+  TableDataType,
+  DatasetSpecificationModel,
+  TableModel,
+  RelationshipModel,
+} from 'generated/tdr';
+import clsx from 'clsx';
+import DatasetSchemaRelationshipModal, { wrapRadioValue } from './DatasetSchemaRelationshipModal';
+import { styles as DatasetCreationStyles } from './DatasetSchemaCommon';
+
+const styles = (theme: CustomTheme) =>
+  ({
+    ...DatasetCreationStyles(theme),
+    schemaSectionHeader: {
+      marginTop: 40,
+      paddingBottom: 5,
+      borderBottom: `2px solid ${theme.palette.terra.green}`,
+      display: 'flex',
+      alignItems: 'center',
+      fontSize: '1.5rem',
+    },
+    schemaSectionHeaderDesc: {
+      marginLeft: 10,
+    },
+    section: {
+      margin: '20px 0 10px',
+    },
+    schemaBuilderContainer: {
+      backgroundColor: theme.palette.primary.focus,
+      padding: 20,
+      marginTop: 20,
+      borderRadius: theme.shape.borderRadius,
+      marginRight: -430,
+    },
+    schemaBuilderMain: {
+      display: 'flex',
+      alignItems: 'flex-start',
+      justifyContent: 'space-between',
+    },
+    schemaBuilderStructureView: {
+      width: '55%',
+      flexShrink: 0,
+    },
+    schemaBuilderStructureViewControls: {
+      display: 'flex',
+      alignItems: 'center',
+      justifyContent: 'space-between',
+      marginTop: 5,
+    },
+    schemaControlButton: {
+      'text-transform': 'none',
+      padding: '10px 14px 10px 11px',
+      marginRight: 15,
+    },
+    iconInButton: {
+      marginRight: 6,
+    },
+    schemaBuilderDetailView: {
+      width: '100%',
+      padding: 10,
+      marginLeft: 10,
+      marginTop: 50,
+    },
+    formInputDatatype: {
+      maxWidth: 200,
+    },
+    jsonWrapper: {
+      backgroundColor: '#272822',
+      borderRadius: 10,
+      padding: 20,
+    },
+  } as any);
+
+interface IProps extends WithStyles<typeof styles> {
+  userEmail: string;
+}
+
+const DatasetSchemaBuilderView = withStyles(styles)(({ classes }: IProps) => {
+  const { register, getValues, setValue } = useFormContext();
+  const [datasetSchema, setDatasetSchema] = useState(
+    getValues().schema as DatasetSpecificationModel,
+  );
+  const [selectedTable, setSelectedTable] = useState(-1);
```

Comment:
Why does this need to default to `-1`? Is that a dummy value to replicate having not started yet? Could undefined be used as a more semantic representation?

Reply:
When I was setting the state to `undefined`, and then changing it, i was getting an error for setting an unset variable - I think this has to stay

Reply:
Interesting! I wonder if that's a typescript thing. Well, sounds fine to me then.